drivers/flash/nrf_qspi_nor.c

```diff
@@ -0,0 +1,688 @@
+	while (size) {
+		if (size == params->size) {
+			/* chip erase */
+			qspi_lock(dev);
+			res = nrfx_qspi_chip_erase();
+			qspi_wait_for_completion(dev);
+			size = 0;
+		} else if ((size >= QSPI_BLOCK_SIZE) &&
+			   QSPI_IS_BLOCK_ALIGNED(addr)) {
+			/* 64 kB block erase */
+			qspi_lock(dev);
+			res = nrfx_qspi_erase(NRF_QSPI_ERASE_LEN_64KB, addr);
+			qspi_wait_for_completion(dev);
+			addr += QSPI_BLOCK_SIZE;
+			size -= QSPI_BLOCK_SIZE;
+		} else if ((size >= QSPI_SECTOR_SIZE) &&
+			   QSPI_IS_SECTOR_ALIGNED(addr)) {
+			/* 4kB sector erase */
+			qspi_lock(dev);
+			res = nrfx_qspi_erase(NRF_QSPI_ERASE_LEN_4KB, addr);
+			qspi_wait_for_completion(dev);
+			addr += QSPI_SECTOR_SIZE;
+			size -= QSPI_SECTOR_SIZE;
+		} else {
+			/* minimal erase size is at least a sector size */
+			LOG_DBG("unsupported at 0x%lx size %zu", (long)addr, size);
+			return -EINVAL;
+		}
+	}
+
+	return qspi_get_zephyr_ret_code(res);
+}
+
+/**
+ * @brief Fills init struct
+ *
+ * @param config Pointer to the config struct provided by user
+ * @param initstruct Pointer to the configuration struct
+ * @retval None
+ */
+static inline void qspi_fill_init_struct(nrfx_qspi_config_t *initstruct)
+{
+	/* Configure XIP offset */
+	initstruct->xip_offset = 0;
+
+	/* Configure pins */
+	initstruct->pins.sck_pin = DT_NORDIC_NRF_QSPI_QSPI_0_SCK_PIN;
+	initstruct->pins.csn_pin = DT_NORDIC_NRF_QSPI_QSPI_0_CSN_PINS_0;
+	initstruct->pins.io0_pin = DT_NORDIC_NRF_QSPI_QSPI_0_IO_PINS_0;
+	initstruct->pins.io1_pin = DT_NORDIC_NRF_QSPI_QSPI_0_IO_PINS_1;
+	initstruct->pins.io2_pin = DT_NORDIC_NRF_QSPI_QSPI_0_IO_PINS_2;
+	initstruct->pins.io3_pin = DT_NORDIC_NRF_QSPI_QSPI_0_IO_PINS_3;
+
+	/* Configure Protocol interface */
+#ifdef DT_INST_0_NORDIC_QSPI_NOR_READOC_ENUM
+	initstruct->prot_if.readoc =
+		(nrf_qspi_writeoc_t)qspi_get_lines_read(DT_INST_0_NORDIC_QSPI_NOR_READOC_ENUM);
+#else
+	initstruct->prot_if.readoc = NRF_QSPI_READOC_FASTREAD;
+#endif
+
+#ifdef DT_INST_0_NORDIC_QSPI_NOR_WRITEOC_ENUM
+	initstruct->prot_if.writeoc =
+		(nrf_qspi_writeoc_t)qspi_get_lines_write(DT_INST_0_NORDIC_QSPI_NOR_WRITEOC_ENUM);
+#else
+	initstruct->prot_if.writeoc = NRF_QSPI_WRITEOC_PP;
+#endif
+	initstruct->prot_if.addrmode =
+		qspi_get_address_size(DT_INST_0_NORDIC_QSPI_NOR_ADDRESS_SIZE_32);
+
+	initstruct->prot_if.dpmconfig = false;
+
+	/* Configure physical interface */
+	initstruct->phy_if.sck_freq =
+		get_nrf_qspi_prescaler(DT_INST_0_NORDIC_QSPI_NOR_SCK_FREQUENCY);
+	initstruct->phy_if.sck_delay = DT_INST_0_NORDIC_QSPI_NOR_SCK_DELAY;
+	initstruct->phy_if.spi_mode = qspi_get_mode(DT_INST_0_NORDIC_QSPI_NOR_CPOL,
+						    DT_INST_0_NORDIC_QSPI_NOR_CPHA);
+
+	initstruct->phy_if.dpmen = false;
+}
+
+/* Configures QSPI memory for the transfer */
+int qspi_nrfx_configure(struct device *dev)
+{
+	if (!dev) {
+		return -ENXIO;
+	}
+
+	/* Main config structure */
+	nrfx_qspi_config_t QSPIconfig;
+
+	qspi_fill_init_struct(&QSPIconfig);
+
+	int res = nrfx_qspi_init(&QSPIconfig, qspi_handler, dev);
```

Comment:
In case this is not successful, the function should no attempt to call `qspi_send_cmd`.

Reply:
Fixed